### PR TITLE
Entity cache: set up basic infrastructure for invalidation

### PR DESCRIPTION
**Pull request by Geal**
This sets up the `Notify` structure to act as pub/sub for invalidation events. A task is started by the entity cache plugin to execute those events. There is apparently no way to mock Redis `SCAN` commands with the `fred` crate, so instead this modifies the integration tests to be able to send back more complex subgraph responses, with headers and extensions. It also adds matching on method, headers, path at the subgraph mock level.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

[ROUTER-380]: https://apollographql.atlassian.net/browse/ROUTER-380?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apollo-router/src/plugins/cache/entity.rs [hunk 1]
```diff
@@ -389,6 +399,15 @@ impl InnerCacheService {
                             }
                         }
 
+                        if let Some(invalidation_extensions) = response
+                            .response
+                            .body_mut()
+                            .extensions
+                            .remove("invalidation")
```
Comment (bnjjj) [inline apollo-router/src/plugins/cache/entity.rs:406]: Could be part of a constant

File: apollo-router/src/plugins/cache/invalidation.rs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    u64_histogram!(
+        "apollo.router.cache.invalidation.keys",
+        "Number of invalidated keys.",
+        count
+    );
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(tag = "kind")]
+pub(crate) enum InvalidationRequest {
```
Comment (bnjjj) [inline apollo-router/src/plugins/cache/invalidation.rs:125]: Do we really want to have the kind field capitalized ?

**Review by bnjjj — APPROVED** [review]

File: apollo-router/src/plugins/cache/invalidation.rs (hunk 2 above)
Comment (Geal) [reply apollo-router/src/plugins/cache/invalidation.rs:125]: do I do that with `#[serde(rename_all = "lowercase")]` over the whole enum?

**Comment by Geal** [comment]
noting that this cannot work right now because redis is only running on linux in CI (I think)

File: apollo-router/src/plugins/cache/invalidation.rs (hunk 2 above)
Comment (Geal) [reply apollo-router/src/plugins/cache/invalidation.rs:125]: done in a189c67bd

**Pull request merged**